src/ast/ast.h

```diff
@@ -615,16 +631,60 @@ class Probe : public Node {
 };
 using ProbeList = std::vector<Probe *>;
 
+class SubprogArg : public Node
+{
+public:
+  DEFINE_ACCEPT
+  DEFINE_LEAFCOPY(SubprogArg)
+
+  SubprogArg(std::string name, SizedType type);
+  ~SubprogArg() = default;
+
+  std::string name() const;
+  SizedType type;
+
+private:
+  SubprogArg(const SubprogArg &other);
+  std::string name_;
+};
+using SubprogArgList = std::vector<SubprogArg *>;
+
+class Subprog : public Scope
+{
+public:
+  DEFINE_ACCEPT
+  DEFINE_LEAFCOPY(Subprog)
+
+  Subprog(std::string name,
+          SizedType type,
+          SubprogArgList *args,
+          StatementList *stmts);
+  ~Subprog();
+
+  SubprogArgList *args = nullptr;
+  SizedType type;
```

Comment:
nit: maybe `return_type` instead of `type`? To differentiate between type of function and return type